---
title: "Simulation study: Part 1"
author: "Stuart Lee"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    fig_caption: yes


vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
```{r knitr.opts, include = FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo=FALSE, 
                      fig.height = 4, fig.width = 6)
```
# Background


# Generating input parameters
We estimate the parameter of a truncated exponential distribution with rate parameter $\theta$ from our filtered Papua New Guinea _P.falciparum_ isolate data using maxiumum likelihood estimation. 
This data contains 104 isolates with approximately 110,000 SNPs. We filter this dataset further to include SNPs that have a minor allele frequency between 0.05 and 0.5. 
```{r load-plot-data, fig.align='center', fig.cap="PNG P.Falciparum minor allele frequency distribution"}
library(haldane)
data(png.allele.frq)

pngMAF <- png.allele.frq$maf
pngMAF.filtered <- pngMAF[pngMAF >= 0.05]

# histogram
hist(pngMAF.filtered, 
     xlim = c(0.05,0.5), 
     xlab = "Minor allele frequency",
     main = "")

```

We want to estimate the parameters of the truncated exponential distribution with support 
$x \in [0.05, 0.5]$. Recall that the probability density function (pdf) $g(x)$ 
for a truncated random variable $X$ on the interval $[a,b]$ is defined as:

\[ g_{X}(x) = \left\{
\begin{array}{l l}
\frac{f(x)}{F(b) - F(a)} & \quad \text{if $x \in [a,b]$}\\
0 & \quad \text{otherwise}
\end{array} \right.\]

where $f_{X}(x)$ and $F_{X}(x)$ are the pdf and cumulative distribution function (cdf) of the non-truncated version of this distribution. For the truncated expontial distribution on the interval $[a,b]$ we obtain the following closed form of the pdf

\[ g_{X}(x) = \frac{\theta\exp(-x\theta)}{\exp(-a\theta) - \exp(-b\theta)}\]

If we consider the sequence $X_{1}, \dots, X_{N}$ of random variables from $\text{TExp}(\theta)$ we can obtain the log-likelihood function as
$$
\begin{aligned}
\log(\mathcal{L}(\theta; X)) &= N\log(\theta) -N\theta\bar{X} -N\log(\exp(-a\theta) - \exp(-b\theta))
\end{aligned}
$$

The MLE, $\hat{\theta}$, is obtained when the first derivative
of the log-likelihood is equal to zero,
$$
\begin{aligned}
\frac{1}{\theta} - \bar{X} - 
(b\exp(-b\theta) + a\exp(a\theta))(\exp(-a\theta) - \exp(-b\theta))^{-1} &= 0
\end{aligned}
$$

The MLE exists if $0 < \bar{X} <= \frac{a + b}{2}$ by Al-Athari (2008), which
is true for our data. 

We solve for the MLE numerically, by finding the root of the first derivative.

```{r log-likelihood, fig.align='center', fig.cap="The density plot of MAF distrubtion with the truncated exponential with MLE for the rate parameter."}
# the log likelihood function for truncated beta
theta.hat <- mleTexp(pngMAF.filtered, 0.05, 0.5)$root

hist(pngMAF.filtered, xlim = c(0.05, 0.5), probability = TRUE, main = "", 
     xlab = "Minor allele frequency")
curve(dtexp(x, theta = theta.hat, lower = 0.05, upper = 0.5), 
      add = TRUE, 
      col = "blue")
save(theta.hat, file = "../../Pf_project/cache/sim_exp_par.rda")
```

The MLE is `r round(theta.hat, 2)`. The fitted distribution doesn't give enough
weight to the left tails. So we try a truncated beta distribution...


