---
title: "Analysis of mixtures of lab strains"
author: "Stuart Lee"
date: "12 April 2016"
output: html_document
---

```{r setup, echo = FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      warning = FALSE, 
                      message = FALSE, 
                      fig.path = "./figures/" )
# libraries 
library(ggplot2)
library(dplyr)

# # functions
# read_coverage <- function(fin) {
#     read.table(fin, header = TRUE, nrow = 1, stringsAsFactors = FALSE)
# }
# 
# read_bs <- function(path) {
#   sample.id <- gsub("(.*\\/)(PG[[:digit:]]{4}-C)(.*)", "\\2", path)
#   read.table(paste0(path, "/fastqc_data_disassemble/_Basic_Statistics.txt"), 
#                        stringsAsFactors = FALSE,
#              header = TRUE) %>% 
#     mutate(sample.id = sample.id)
# }
# 
# read_bq <- function(path) {
#   sample.id <- gsub("(.*\\/)(PG[[:digit:]]{4}-C)(.*)", "\\2", path)
# 
#   read.table(paste0(path, "/fastqc_data_disassemble/_Per_base_sequence_quality_Mean.txt"),
#              stringsAsFactors = FALSE,
#              header = TRUE) %>% mutate(sample.id = sample.id)
# }
# 
# read_gc <- function(path) {
#   sample.id <- gsub("(.*\\/)(PG[[:digit:]]{4}-C)(.*)", "\\2", path)
# 
#   g <- read.table(paste0(path, "/fastqc_data_disassemble/_Per_base_sequence_content_C.txt"), 
#                   stringsAsFactors = FALSE,
#                   header = TRUE)
#   c <- read.table(paste0(path, "/fastqc_data_disassemble/_Per_base_sequence_content_G.txt"), 
#                   stringsAsFactors = FALSE,
#                   header = TRUE)
#   gc <- left_join(c, g) %>% mutate(gc_content = G+C, sample.id = sample.id)
#   gc
# }

```
# Introduction

We have analysed 27 mixture isolates from lab-strains obtained from the 
[Malaria Genomics Consortium](https://www.malariagen.net/data/pf3k-5) pilot
data release version 5. These samples contain mixtures of lab-strains, 3D7,
Dd2, HB3, and 7G8 in varying proportions. There are at most three mixtures of
different strains. They were sequenced by Dr Jason Wendler
at the Wellcome Trust Sanger Institute and can be found at the European Nucleotide
Archive under study accession [PRJEB4338](http://www.ebi.ac.uk/ena/data/view/PRJEB4338). 

```{r metadata}
sample_info <- read.table("../metadata/pf3k_release_5_mixtures_metadata.txt",
                          header = TRUE,
                          stringsAsFactors = FALSE) %>% 
  mutate(MOI = ((X3D7>0)+(Dd2>0)+(HB3>0)+(X7G8>0)))


knitr::kable(sample_info %>% 
                 select(sample, MOI, X3D7, Dd2, HB3, X7G8),
             caption = "Sample MOI and mixture proportions.")

```
# Assessing clonality using biallelic SNPs

We have performed analysed the mixture isolates by combining variant calls obtained
from the Malaria Genomics Consortium SNP calling pipeline. This pipeline was run
on 2,640 isolates, which includes the 27 mixture isolates and 98 lab-strain isolates
used for genetic cross experiment. We combined the vcf files from 14 nuclear 
_P.falciparum_ chromosomes for
all isolates and then filtered based on the following criteria:

* extract variants that are on Core regions of the genome
* extract variants with a log-odds Variant Quality Score > 0
* keep only biallelic SNPs

After the initial filtering we then removed the 125 lab-strain isolates
to extract a set of globally polymorphic alleles by filtering sites with
minor allele frequencies less than 5 per cent. Then we took the list of
approximately 19,000 globally informative SNPs and extracted variant call
information at those positions for the 27 mixture isolates.

We then converted this file to GDS format after adjusting the AD tag header to
meet the VCF4.2 specifications.

We use `moimix` to estimate the number and proportion of mixed infections in each sample and
`estmoi` with default settings to estimate the number of clones. 

# Initial QC

```{r read_gds}
library(SeqArray)
library(moimix)
isolates <- seqOpen("../processed_data/mixtures_only.gds")

coords <- getCoordinates(isolates)

gatk_df <- getGATKInfo(isolates)
sample.id <- seqGetData(isolates, "sample.id")

knitr::kable(count(coords, chromosome))

counts_matrix <- alleleCounts(isolates)

snp_coverage <- coverageBySample(counts_matrix, 20)

plot(ecdf(snp_coverage), 
     main = "ECDF of proportion of samples with SNVs covered up to 20x")

seqSetFilter(isolates, 
             variant.id = as.integer(names(snp_coverage[snp_coverage > 0.9])))

```

We see in the ECDF plot that the most variants are highly covered. 
We apply a further filter by only keeping variants where at least 90\% of the 
samples are covered to at least 20 bases.
This reduces the call-set to `r length(seqGetData(isolates, "variant.id"))` variants. 
For each sample we plot the histogram and spectra of the estimated non-reference allele 
frequencies genome-wide. 

```{r, baf_matrix, fig.width=8, fig.height=4}
baf_obj <- bafMatrix(isolates)

for(id in sample.id) {
  par(mfrow = c(1,2))
  title <- paste("ID 3D7 Dd2 HB3 7G8\n",
                 paste(select(filter(sample_info, sample == id), 
                              sample, X3D7, Dd2, HB3, X7G8), collapse = " "))
  hist(baf_obj$baf_matrix[id, ], breaks = 100, xlab = "NRAF", main = title )
  plot(baf_obj, id)
}

```


Since we have alinged to the 3D7 reference genome, in cases where there are
mixtures of 3D7 with another strain, the NRAF produces a single band centred at the
non-reference strains proportion, i.e. we are variant in non-reference strain but
not to the reference genome itself. For mixtures of three isolates we produce 3
distinct clusters. We use `moimix` function `binommix` to fit our binomial
mixture models, trying k = 1 to 5 clones.

We then use an error metric to choose the number of components to choose, that is,
we select components based on the choice of K that minmises the between the absolute
distance between the the NRAF and the estimated parameters

```{r, construct_models, cache=TRUE}
#seqSetFilter(isolates, 
#             variant.id = gatk_df$variant.id[gatk_df$qd > 30])

tryK <- 1:5
models <- list()
counts_matrix <- alleleCounts(isolates)
set.seed(1101)
for(id in sample.id) {
  models[[id]] <- binommix(counts_matrix, id, tryK, coverage_threshold = 20L)
  break
}

moi_models <- lapply(models, function(m) getModel(m$fits, "BIC") )
moi_estimated_bic <- lapply(moi_models, function(m) m@k) %>% unlist
moi_estimated_param <- lapply(moi_models, getTheta) %>% bind_rows
```

We also ran `estMOI` with the default settings, considering SNP trios to form
haplotypes and using a 90th percentile cut-off for the genome-wide estimate of 
MOI.

```{r, parseEstMOI}
estmoi_data <- readEstMOI("../processed_data/moi_results/")
estmoi_estimates <- filter(estmoi_data, estimate == "MOI-estimate") %>% 
  select(sample.id[7], estMOI = K, proportion, count)


```

Finally, we estimated the $F_{ws}$ statistic for each isolate using the method
proposed by Manske et al., 2012.

```{r fws}

fws_all <- getFws(isolates)

fws_dat <- data.frame(sample= names(fws_all), fws = fws_all)

```

To compare each method we need compare each predicted estimate of the number
of clones to the true multiplicity of infeciton obtained by `estmoi` and
`moimix`.

```{r compare_moi}
moi_estimated_df <- data.frame(sample = names(moi_estimated_bic), 
                               moimix_K = moi_estimated_bic)

all_moi <- left_join(left_join(left_join(sample_info, estmoi_estimates,
                     by = c("sample" = "sample.id")),
                     fws_dat), moi_estimated_df)
knitr::kable(all_moi)
```

