# Title: coil_functions.R
# Author: Stuart Lee
# Description: Helper functions for extracting barcodes
# for use in COIL - see 
# http://www.broadinstitute.org/infect/malaria/coil/

#' Extract barcode for use in COIL program
#' 
#' @param gdsfile a \code{\link[SeqArray]{SeqVarGDSClass}} object
#' @param variant.id a numeric vector of variant identifiers
#' @param barcode.file character name of output file
#' 
#' @details Generate SNP barcode using supplied variant ids. This
#' is constructed by collapsing alleles over the variant ids, forming
#' a barcode for each sample. Then a barcode file containing
#' sample ids + barcode is generated and can be supplied to the COIL web-app
#' or to the command line COIL program. 
#' @references Galinsky, Kevin, et al. "COIL: a methodology for evaluating 
#' malarial complexity of infection using likelihood from single nucleotide 
#' polymorphism data." Malaria journal 14.1 (2015): 4. 
#' @note Need to look at how COIL constructed it's orginal barcode. Esp. het calls.
#' @return data.frame containing barcode for each sample
#' @importFrom SeqArray seqSetFilter seqGetData
#' @export
extractBarcode <- function(gdsfile, variant.id, barcode.file) {
    stopifnot(inherits(gdsfile, "SeqVarGDSClass"))
    
    old.filter <- seqGetFilter(gdsfile)
    seqSetFilter(gdsfile, variant.id = variant.id)
    
    sample.id <- seqGetData(gdsfile, "sample.id")
    major.alleles <- callMajor(gdsfile, get.nucleotides = TRUE)
    # paste haplotypes together
    barcode <- apply(major.alleles, 1, paste, collapse = "")
    # add in sample.id
    stopifnot(length(sample.id) == length(barcode))
    final_barcode <- data.frame(sample.id, barcode)
    # write out barcode file
    datafile <- file(barcode.file, open = "wt")
    on.exit(close(datafile))
    # construct header and write it out
    header <- paste0("# COIL barcode generated by moimix ", Sys.time(), "\n",
                     "#ID Barcode")
    writeLines(header, con = datafile)
    write.table(final_barcode, datafile, col.names = FALSE, row.names = FALSE, quote = FALSE)
    
    # reset filter
    seqSetFilter(gdsfile, sample.id = sample.id, variant.sel = old.filter$variant.sel)
    # return the final barcode
    invisible(final_barcode)
}