#' Simple simulation of MOI SNV data
#' 
#' @description Generate read count data in support of SNVs partioned by 
#' mixtures of clones within a host. This is done by setting true
#' clonal mixture proportions and the underlying sampling probabilities
#' for each clone. Then a random sample of underlying SNV probabilities
#' is sampled from either a rbeta or rtexp distrubition with given parameters
#' and read counts are generated by bionimial conditonal on the underlying
#' SNV probability with the clone sampling probabilities. 
#' 
#' @param n.samples number of infected hosts in population
#' @param n.snps number of SNPs observed
#' @param moi number of clonal infections
#' @param coverage vector of total coverage per sample
#' @param clonal.props optional matrix of true mixture proportions
#' @param geno.props optional matrix of true mixture components 
#' @param aaf.dist  sampling distribution for SNV frequencies
#' @param ... other parameters to pass to aaf.dist
#' @return A list containing the following elements
#'  pi.true an n.samples by moi matrix containing true clonal proportions
#'  mu.true an n.samplps by moi matrix containing true genotype proportions
#'  aaf an n.snps vector of under SNV probabilities
#'  reads an n.samples by n.snps matrix containing read counts supporting each SNV 
#' @importFrom MCMCpack rdirichlet
#' @import foreach
#' @export

simulate_moi <- function(n.samples, n.snps, moi, coverage, 
                         pi.true = NULL, mu.true = NULL, 
                         aaf.dist, ...) {
    # I/0 checking
    #-- todo
    # step 1, generate underlying parameters
    if (is.null(pi.true)) {
        pi.true <- t(MCMCpack::rdirichlet(n.samples, 
                                          alpha = rep(1, moi)))
        
    }
    
    if (is.null(mu.true)) {
        # produces an moi by n.samples
        mu.true <- replicate(n.samples, runif(moi))
    }
    print(dim(pi.true)); print(pi.true)
    print(dim(mu.true)); print(mu.true)
    # generate mixture indexes for each SNV for each isolate
    # produce an n.samples by n.snps matrix with assignments
    clusters <- foreach(i=1:n.samples, .combine = cbind) %do% {
        sample(1:moi, n.snps, replace = TRUE, prob = pi.true[,i])
    } 
    print(dim(clusters)); print(clusters)
    # generate underlying SNV frequenciences
    aaf.dist <- match.fun(aaf.dist)
    aaf <- aaf.dist(n=n.snps, ...)
    # conditional probabilities of observing each SNV given underlying
    # clonal genotype
    # generate matrix of assignments
    sample.p <- foreach(i =1:n.samples, .combine = cbind) %do% {
        mu.true[clusters[,i],i]
    }
    # multiply by the frequency of each alternate allele 
    sample.p <- sample.p * aaf
    # generate observed read counts in support of each SNV for each clone
    read.counts <- foreach(i = 1:n.samples, .combine = cbind) %do% {
        rbinom(n.snps, size = coverage, prob = sample.p[,i])
    }
    # generate errors -- TODO
    return(list(pi.true = pi.true, 
                mu.true = mu.true,
                
                read.counts = read.counts))
}