#' Simple simulation of MOI SNV data
#' 
#' @description Generate read count data in support of SNVs partioned by 
#' mixtures of clones within a host. This is done by setting true
#' clonal mixture proportions and the underlying sampling probabilities
#' for each clone. Then a random sample of underlying SNV probabilities
#' is sampled from either a rbeta or rtexp distrubition with given parameters
#' and read counts are generated by bionimial conditonal on the underlying
#' SNV probability with the clone sampling probabilities. 
#' 
#' @param n.samples number of infected hosts in population
#' @param n.snps number of SNPs observed
#' @param moi number of clonal infections
#' @param coverage vector of total coverage per sample
#' @param clonal.props optional matrix of true mixture proportions
#' @param geno.props optional matrix of true mixture components 
#' @param aaf.dist  sampling distribution for SNV frequencies
#' @param ... other parameters to pass to aaf.dist
#' @return A list containing the following elements
#'  pi.true an n.samples by moi matrix containing true clonal proportions
#'  mu.true an n.samplps by moi matrix containing true genotype proportions
#'  aaf an n.snps vector of under SNV probabilities
#'  reads an n.samples by n.snps matrix containing read counts supporting each SNV 
#' @importFrom MCMCpack rdirichlet
#' 
#' @export

simulate_moi <- function(n.samples, n.snps, moi, coverage, 
                         pi.true = NULL, mu.true = NULL, 
                         aaf.dist, ...) {
    # I/0 checking
    #-- todo
    # step 1, generate underlying parameters
    if (is.null(pi.true)) {
        pi.true <- MCMCpack::rdirichlet(n.samples, alpha = rep(1, moi))
        
    }
    
    if (is.null(mu.true)) {
        # produces an moi by n.samples
        mu.true <- replicate(n.samples, runif(moi))
    }
    
    # generate mixture indexes for each SNV for each isolate
    # produce an n.samples by n.snps matrix with assignments
    clusters <- apply(pi.true, 1, 
                      function(p) sample(1:moi, 
                                         n.snps, 
                                         replace = TRUE, 
                                         prob = p)) 
    # generate underlying SNV frequenciences
    aaf.dist <- match.fun(aaf.dist)
    aaf <- aaf.dist(n=n.snps, ...)
    # conditional probabilities of observing each SNV given underlying
    # clonal genotype
    
}